slatedb/src/transactional_object/mod.rs

```diff
@@ -0,0 +1,919 @@
+        self.id == *other
+    }
+}
+
+impl PartialEq<MonotonicId> for u64 {
+    fn eq(&self, other: &MonotonicId) -> bool {
+        *self == other.id
+    }
+}
+
+impl std::ops::Add<u64> for MonotonicId {
+    type Output = MonotonicId;
+
+    fn add(self, rhs: u64) -> Self::Output {
+        MonotonicId::new(self.id + rhs)
+    }
+}
+
+impl MonotonicId {
+    pub(crate) fn initial() -> Self {
+        Self { id: 1 }
+    }
+
+    pub(crate) fn new(id: u64) -> Self {
+        Self { id }
+    }
+
+    pub(crate) fn id(&self) -> u64 {
+        self.id
+    }
+
+    pub(crate) fn next(&self) -> Self {
+        Self { id: self.id + 1 }
+    }
+}
+
+impl From<u64> for MonotonicId {
+    fn from(id: u64) -> Self {
+        Self::new(id)
+    }
+}
+
+impl From<MonotonicId> for u64 {
+    fn from(id: MonotonicId) -> Self {
+        id.id
+    }
+}
+
+/// Generic file metadata for versioned objects
+#[derive(Debug)]
+pub(crate) struct GenericObjectMetadata<Id: Copy = MonotonicId> {
+    pub(crate) id: Id,
+    pub(crate) location: Path,
+    pub(crate) last_modified: chrono::DateTime<Utc>,
+    #[allow(dead_code)]
+    pub(crate) size: u32,
+}
+
+/// A local view of a transactional object, possibly with local mutations
+#[derive(Clone, Debug)]
+pub(crate) struct DirtyObject<T, Id: Copy = MonotonicId> {
+    /// This ID of the object from which this `DirtyObject` was created
+    id: Id,
+    pub(crate) value: T,
+}
+
+impl<T, Id: Copy> DirtyObject<T, Id> {
+    #[allow(dead_code)]
+    pub(crate) fn id(&self) -> Id {
+        self.id
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn into_value(self) -> T {
+        self.value
+    }
+}
+
+/// An in-memory datum that is backed by durable storage and can be
+/// transactionally updated.
+#[async_trait::async_trait]
+pub(crate) trait TransactionalObject<T: Clone, Id: Copy = MonotonicId> {
+    /// Returns the version ID of the in-memory view of the object
+    fn id(&self) -> Id;
+
+    /// Returns the in-memory view of the object
+    fn object(&self) -> &T;
+
+    /// Returns a `DirtyObject` with the current version ID and object which can be
+    /// modified locally and passed to `update` to persist mutations durably.
+    fn prepare_dirty(&self) -> Result<DirtyObject<T, Id>, TransactionalObjectError>;
+
+    /// Refreshes the in-memory view of the object with the latest version stored durably.
+    /// This may result in a different in-memory view returned by `object` (and different id
+    /// returned by `id`) if another process has successfully updated the object.
+    async fn refresh(&mut self) -> Result<&T, TransactionalObjectError>;
+
+    /// Transactionally update the object. Will succeed iff the version id in durable storage
+    /// matches the version id of the provided `DirtyObject`. If the versions don't match
+    /// then this fn returns `ObjectVersionExists`.
+    async fn update(&mut self, dirty: DirtyObject<T, Id>) -> Result<(), TransactionalObjectError>;
```

Comment:
this API feels a little weird to me - maybe it's just a rust-ism, but I'd prefer something like:
```rs
let dirty = txn_object.prepare_dirty();
dirty.value = ...;
dirty.commit();
```
And ideally, once I call `prepare_dirty()` it "consumes" the `txn_object` so I can't called `prepare_dirty` twice on a single `txn_object` (that would never, succeed, right?)